File: leetcode/number-of-islands/solution2.cpp
// Union-Find. O(rows * cols * a(n)) time, O(rows * cols) space.
// Does not mutate the grid, and extends to "Number of Islands II" where land
// arrives one cell at a time and the count is queried after each addition.
class Solution {
public:
    vector<int> parent, rnk;

    int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];  // path halving
            x = parent[x];
        }
        return x;
    }

    bool unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (rnk[ra] < rnk[rb]) swap(ra, rb);
        parent[rb] = ra;
        if (rnk[ra] == rnk[rb]) rnk[ra]++;
        return true;
    }

    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty() || grid[0].empty()) return 0;

        int rows = grid.size(), cols = grid[0].size();
        parent.resize(rows * cols);
        iota(parent.begin(), parent.end(), 0);
        rnk.assign(rows * cols, 0);

        int islands = 0;

        // Only up and left: the down/right pairs get handled from the other side.
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] != '1') continue;

                islands++;
                int id = i * cols + j;
                if (i > 0 && grid[i - 1][j] == '1' && unite(id - cols, id)) islands--;
                if (j > 0 && grid[i][j - 1] == '1' && unite(id - 1, id)) islands--;
            }
        }

        return islands;
    }
};

File: leetcode/number-of-islands/solution2.py
from typing import List


class DSU:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]  # path halving
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.rank[ra] < self.rank[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        if self.rank[ra] == self.rank[rb]:
            self.rank[ra] += 1
        return True


class Solution:
    """Union-Find. O(rows * cols * a(n)) time, O(rows * cols) space.

    Slower than flood fill in practice and it does not mutate the grid, but it
    is the approach that extends to "Number of Islands II" (land added one cell
    at a time, report the count after each addition) — flood fill has to redo
    the whole sweep per query, union-find just unions the new cell's neighbours.

    Only the up and left neighbours are scanned: by the time cell (i, j) is
    reached, the down/right pairs will be handled from the other side.
    """

    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        rows, cols = len(grid), len(grid[0])
        dsu = DSU(rows * cols)
        islands = 0

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] != "1":
                    continue

                islands += 1
                if i > 0 and grid[i - 1][j] == "1" and dsu.union((i - 1) * cols + j, i * cols + j):
                    islands -= 1
                if j > 0 and grid[i][j - 1] == "1" and dsu.union(i * cols + j - 1, i * cols + j):
                    islands -= 1

        return islands
